### PR TITLE
Implement a wrapper for 'docker inspect' command

### PR DESCRIPTION
**Pull request by GusAntoniassi**
Solves #489, feedback is highly appreciated.

### CODE REVIEW

File: modules/docker/inspect.go [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+			HostIp   string
+			HostPort string
+		}
+	}
+	HostConfig struct {
+		Binds []string
+	}
+}
+
+func Inspect(t *testing.T, id string) ContainerInspect {
```
Comment (brikis98) [inline modules/docker/inspect.go:82]: NIT: Please add comment using the standard form for all functions in Terratest

File: modules/docker/inspect.go [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+	return out
+}
+
+// InspectE runs the 'docker inspect {container id}' command and returns a ContainerInspect
+// struct, converted from the output JSON, along with any errors
+func InspectE(t *testing.T, id string) (ContainerInspect, error) {
+	match, _ := regexp.MatchString("[[:alnum:]]", id)
+	if !match {
+		return ContainerInspect{}, fmt.Errorf("%s is not a valid container ID", id)
+	}
```
Comment (brikis98) [inline modules/docker/inspect.go:94]: NIT: This is a nice idea, but won't the `docker inspect` command itself give a reasonable error if it's an invalid ID?

File: modules/docker/inspect.go [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+
+func Inspect(t *testing.T, id string) ContainerInspect {
+	out, err := InspectE(t, id)
+	require.NoError(t, err)
+	return out
+}
+
+// InspectE runs the 'docker inspect {container id}' command and returns a ContainerInspect
+// struct, converted from the output JSON, along with any errors
+func InspectE(t *testing.T, id string) (ContainerInspect, error) {
```
Comment (brikis98) [inline modules/docker/inspect.go:90]: NIT: Consider returning `*ContainerInspect`, so if there's an error, you can return `nil, err`.

File: modules/docker/inspect.go [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+		return ContainerInspect{}, err
+	}
+
+	var containers []inspectOutput
+	err = json.Unmarshal([]byte(out), &containers)
+	if err != nil {
+		return ContainerInspect{}, err
+	}
+
+	if len(containers) == 0 {
```
Comment (brikis98) [inline modules/docker/inspect.go:112]: Should that be an error?

File: modules/docker/inspect.go [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+		return ContainerInspect{}, nil
+	}
+
+	c := containers[0]
```
Comment (brikis98) [inline modules/docker/inspect.go:116]: NIT: Please minimize use of single-letter variable names (here and later in the code), as they are hard to follow.

File: modules/docker/inspect.go [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+}
+
+// transformContainerPorts converts 'docker inspect' output JSON into a more friendly and testable format
+func transformContainer(t *testing.T, c inspectOutput) ContainerInspect {
+	name := strings.TrimLeft(c.Name, "/")
+	ports := transformContainerPorts(t, c)
+	volumes := transformContainerVolumes(t, c)
+
+	created, err := time.Parse(time.RFC3339Nano, c.Created)
+	require.NoError(t, err)
```
Comment (brikis98) [inline modules/docker/inspect.go:128]: In order to fulfill the contract of `E` functions (e.g., `inspectE`), this function should return the error rather than calling `require.NoError`.

File: modules/docker/inspect.go [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+func transformContainerPorts(t *testing.T, c inspectOutput) []Port {
+	var ports []Port
+
+	cPorts := c.NetworkSettings.Ports
+
+	for k, pb := range cPorts {
+		split := strings.Split(k, "/")
+
+		containerPort, err := strconv.ParseUint(split[0], 10, 16)
+		require.NoError(t, err)
```
Comment (brikis98) [inline modules/docker/inspect.go:163]: Same here and below. Return errors, rather than calling `require.NoError`.

File: modules/docker/inspect.go [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+
+		protocol := split[1]
```
Comment (brikis98) [inline modules/docker/inspect.go:165]: Check `len(split) > 1` first!

File: modules/docker/inspect.go [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+
+		for _, p := range pb {
```
Comment (brikis98) [inline modules/docker/inspect.go:167]: NIT: Two letter variable names are also tough to follow. I had to keep scrolling up and down to remember what `p` and `pb` were.

File: modules/docker/inspect.go [hunk 10]
```diff
@@ -0,0 +1,204 @@
...
+	volumes := make([]VolumeBind, 0, len(binds))
+
+	for _, b := range binds {
+		var source, dest string
+
+		split := strings.Split(b, ":")
+
+		// Considering it a named volume
+		source = split[0]
+		dest = split[1]
```
Comment (brikis98) [inline modules/docker/inspect.go:195]: Check `len(split) == 2`!

File: modules/docker/inspect_test.go [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+	"time"
+)
+
+const image = "nginx:1.17-alpine"
+
+func TestInspect(t *testing.T) {
+	t.Parallel()
+
+	// append timestamp to container name to allow running tests in parallel
+	name := "inspect-test-" + strconv.FormatInt(time.Now().UnixNano(), 10)
```
Comment (brikis98) [inline modules/docker/inspect_test.go:18]: You can use `random.UniqueId()` from Terratest instead

File: modules/docker/inspect_test.go [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+import (
+	"fmt"
+	"github.com/gruntwork-io/terratest/modules/shell"
+	"github.com/stretchr/testify/require"
+	"strconv"
+	"testing"
+	"time"
+)
+
+const image = "nginx:1.17-alpine"
```
Comment (brikis98) [inline modules/docker/inspect_test.go:12]: Could you give this a more descriptive name? E.g., `dockerImageForInspectTest`.

File: modules/docker/inspect_test.go [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+
+func TestInspect(t *testing.T) {
```
Comment (brikis98) [inline modules/docker/inspect_test.go:14]: Thanks for adding tests! 👍

File: modules/docker/inspect_test.go [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+func TestInspectWithInvalidContainerID(t *testing.T) {
+	_, err := InspectE(t, "This is not a valid container ID")
+	require.Error(t, err)
+}
+
+func removeContainer(t *testing.T, id string) {
+	cmd := shell.Command{
+		Command: "docker",
+		Args:    []string{"container", "rm", "-f", id},
+	}
```
Comment (brikis98) [inline modules/docker/inspect_test.go:85]: Do you need to run `docker stop` first?

File: modules/docker/inspect.go (hunk 2 above)
Comment (GusAntoniassi) [reply modules/docker/inspect.go:94]: Yep, now I checked and it returns `Error: No such container: This is not a valid container ID`, which is pretty straightforward. Gonna remove this validation.

File: modules/docker/inspect.go (hunk 8 above)
Comment (GusAntoniassi) [reply modules/docker/inspect.go:165]: From what I have tested Docker won't return a `HostConfig.Binds` without a colon, but I'm not 100% sure on the internals so it makes sense to add a check just to be safe.

Also gonna add tests for named and anonymous volumes, sounds safer that way.

File: modules/docker/inspect.go (hunk 10 above)
Comment (GusAntoniassi) [reply modules/docker/inspect.go:195]: From what I have tested Docker won't return a HostConfig.Binds without a colon, but I'm not 100% sure on the internals so it makes sense to add a check just to be safe.

Also gonna add tests for named and anonymous volumes, sounds safer that way.

File: modules/docker/inspect.go (hunk 4 above)
Comment (GusAntoniassi) [reply modules/docker/inspect.go:112]: Good point. 

From what I have tested Docker returns `Error: No such container: abcde123456`, so I'm removing this check and adding a test case for a valid but unknown container ID.

File: modules/docker/inspect_test.go (hunk 14 above)
Comment (GusAntoniassi) [reply modules/docker/inspect_test.go:85]: Not in this case, since I'm passing the `-f` flag Docker will remove it even if it is running

Comment (zackproser) [reply modules/docker/inspect_test.go:85]: I would vote for the more verbose `--force` since, as you suggest, it will be easier to understand quickly what's happening.

Comment (GusAntoniassi) [reply modules/docker/inspect_test.go:112]: Makes sense @zackproser, changed it to `--force` now, thanks for the input!

**Comment by GusAntoniassi** [comment]
Thanks for the review, @brikis98

File: modules/docker/inspect.go [hunk 15]
```diff
@@ -0,0 +1,214 @@
...
+		return nil, err
+	}
+
+	var containers []inspectOutput
+	err = json.Unmarshal([]byte(out), &containers)
+	if err != nil {
+		return nil, err
+	}
+
+	container := containers[0]
```
Comment (brikis98) [inline modules/docker/inspect.go:108]: Probably best to check `len(containers) > 0`. Not sure it's likely to happen, but better to check than panic at test time.

File: modules/docker/inspect_test.go [hunk 16]
```diff
@@ -0,0 +1,115 @@
...
+
+func TestInspectWithExposedPort(t *testing.T) {
+	t.Parallel()
+
+	// choosing an unique high port to avoid conflict on test machines
+	port := 13031
+
+	options := &RunOptions{
+		Detach: true,
+		OtherOptions:         []string{fmt.Sprintf("-p=%d:80", port)},
```
Comment (brikis98) [inline modules/docker/inspect_test.go:45]: Indentation looks off here. Make sure to run `gofmt` / `goimports`.

File: modules/docker/inspect.go (hunk 15 above)
Comment (GusAntoniassi) [reply modules/docker/inspect.go:113]: True, by removing the check I accidentally removed a safe check for index out of range. I've added it back, returning an error it `containers` is empty.

**Comment by brikis98** [comment]
OK, great, thanks! Kicking off tests now.

**Comment by brikis98** [comment]
The `goimports` pre-commit check failed. Please be sure to run `goimports` on the code!

**Comment by GusAntoniassi** [comment]
Sorry @brikis98, I read about the pre-commit in the contributing guide but forgot to execute it.

The imports should be fixed now.

**Comment by yorinasub17** [comment]
Thanks just kicked off another build.

**Comment by brikis98** [comment]
All the new inspect tests failed with similar errors:

```
=== RUN   TestInspect
=== PAUSE TestInspect
=== CONT  TestInspect
TestInspect 2020-04-09T06:53:50Z run.go:62: Running 'docker run' on image 'nginx:1.17-alpine'
TestInspect 2020-04-09T06:53:50Z command.go:87: Running command docker with args [run --detach --name inspect-test-bsN5FJ nginx:1.17-alpine]
TestInspect 2020-04-09T06:53:51Z command.go:158: Unable to find image 'nginx:1.17-alpine' locally
TestInspect 2020-04-09T06:53:51Z command.go:158: 1.17-alpine: Pulling from library/nginx
TestInspect 2020-04-09T06:53:52Z command.go:158: 4167d3e14976: Pulling fs layer
TestInspect 2020-04-09T06:53:52Z command.go:158: bb292c78f105: Pulling fs layer
TestInspect 2020-04-09T06:53:52Z command.go:158: 4167d3e14976: Verifying Checksum
TestInspect 2020-04-09T06:53:52Z command.go:158: 4167d3e14976: Download complete
TestInspect 2020-04-09T06:53:52Z command.go:158: bb292c78f105: Verifying Checksum
TestInspect 2020-04-09T06:53:52Z command.go:158: bb292c78f105: Download complete
TestInspect 2020-04-09T06:53:52Z command.go:158: 4167d3e14976: Pull complete
TestInspect 2020-04-09T06:53:53Z command.go:158: bb292c78f105: Pull complete
TestInspect 2020-04-09T06:53:53Z command.go:158: Digest: sha256:abe5ce652eb78d9c793df34453fddde12bb4d93d9fbf2c363d0992726e4d2cad
TestInspect 2020-04-09T06:53:53Z command.go:158: Status: Downloaded newer image for nginx:1.17-alpine
TestInspect 2020-04-09T06:53:54Z command.go:158: 64e2a233bec6d5487b7b1ba4cb4ca3e73703eb9ef93f2935d9f22dcab0f0fa22
TestInspect 2020-04-09T06:53:56Z command.go:87: Running command docker with args [container inspect Unable to find image 'nginx:1.17-alpine' locally
1.17-alpine: Pulling from library/nginx
4167d3e14976: Pulling fs layer
bb292c78f105: Pulling fs layer
4167d3e14976: Verifying Checksum
4167d3e14976: Download complete
bb292c78f105: Verifying Checksum
bb292c78f105: Download complete
4167d3e14976: Pull complete
bb292c78f105: Pull complete
Digest: sha256:abe5ce652eb78d9c793df34453fddde12bb4d93d9fbf2c363d0992726e4d2cad
Status: Downloaded newer image for nginx:1.17-alpine
64e2a233bec6d5487b7b1ba4cb4ca3e73703eb9ef93f2935d9f22dcab0f0fa22]
TestInspect 2020-04-09T06:53:56Z command.go:158: Error: No such container: Unable to find image 'nginx:1.17-alpine' locally
TestInspect 2020-04-09T06:53:56Z command.go:158: 1.17-alpine: Pulling from library/nginx
TestInspect 2020-04-09T06:53:56Z command.go:158: 4167d3e14976: Pulling fs layer
TestInspect 2020-04-09T06:53:56Z command.go:158: bb292c78f105: Pulling fs layer
TestInspect 2020-04-09T06:53:56Z command.go:158: 4167d3e14976: Verifying Checksum
TestInspect 2020-04-09T06:53:56Z command.go:158: 4167d3e14976: Download complete
TestInspect 2020-04-09T06:53:56Z command.go:158: bb292c78f105: Verifying Checksum
TestInspect 2020-04-09T06:53:56Z command.go:158: bb292c78f105: Download complete
TestInspect 2020-04-09T06:53:56Z command.go:158: 4167d3e14976: Pull complete
TestInspect 2020-04-09T06:53:56Z command.go:158: bb292c78f105: Pull complete
TestInspect 2020-04-09T06:53:56Z command.go:158: Digest: sha256:abe5ce652eb78d9c793df34453fddde12bb4d93d9fbf2c363d0992726e4d2cad
TestInspect 2020-04-09T06:53:56Z command.go:158: Status: Downloaded newer image for nginx:1.17-alpine
TestInspect 2020-04-09T06:53:56Z command.go:158: 64e2a233bec6d5487b7b1ba4cb4ca3e73703eb9ef93f2935d9f22dcab0f0fa22
TestInspect 2020-04-09T06:53:56Z command.go:158: []
TestInspect 2020-04-09T06:53:56Z command.go:87: Running command docker with args [container rm --force Unable to find image 'nginx:1.17-alpine' locally
1.17-alpine: Pulling from library/nginx
4167d3e14976: Pulling fs layer
bb292c78f105: Pulling fs layer
4167d3e14976: Verifying Checksum
4167d3e14976: Download complete
bb292c78f105: Verifying Checksum
bb292c78f105: Download complete
4167d3e14976: Pull complete
bb292c78f105: Pull complete
Digest: sha256:abe5ce652eb78d9c793df34453fddde12bb4d93d9fbf2c363d0992726e4d2cad
Status: Downloaded newer image for nginx:1.17-alpine
64e2a233bec6d5487b7b1ba4cb4ca3e73703eb9ef93f2935d9f22dcab0f0fa22]
TestInspect 2020-04-09T06:53:56Z command.go:158: Error: No such container: Unable to find image 'nginx:1.17-alpine' locally
TestInspect 2020-04-09T06:53:56Z command.go:158: 1.17-alpine: Pulling from library/nginx
TestInspect 2020-04-09T06:53:56Z command.go:158: 4167d3e14976: Pulling fs layer
TestInspect 2020-04-09T06:53:56Z command.go:158: bb292c78f105: Pulling fs layer
TestInspect 2020-04-09T06:53:56Z command.go:158: 4167d3e14976: Verifying Checksum
TestInspect 2020-04-09T06:53:56Z command.go:158: 4167d3e14976: Download complete
TestInspect 2020-04-09T06:53:56Z command.go:158: bb292c78f105: Verifying Checksum
TestInspect 2020-04-09T06:53:56Z command.go:158: bb292c78f105: Download complete
TestInspect 2020-04-09T06:53:56Z command.go:158: 4167d3e14976: Pull complete
TestInspect 2020-04-09T06:53:56Z command.go:158: bb292c78f105: Pull complete
TestInspect 2020-04-09T06:53:56Z command.go:158: Digest: sha256:abe5ce652eb78d9c793df34453fddde12bb4d93d9fbf2c363d0992726e4d2cad
TestInspect 2020-04-09T06:53:56Z command.go:158: Status: Downloaded newer image for nginx:1.17-alpine
TestInspect 2020-04-09T06:53:56Z command.go:158: 64e2a233bec6d5487b7b1ba4cb4ca3e73703eb9ef93f2935d9f22dcab0f0fa22
--- FAIL: TestInspect (5.83s)
--- FAIL: TestInspect (5.83s)
    inspect.go:86: 
        	Error Trace:	inspect.go:86
        	            				inspect_test.go:30
        	Error:      	Received unexpected error:
        	            	exit status 1
        	Test:       	TestInspect
    command.go:33: exit status 1
```

**Comment by GusAntoniassi** [comment]
That's odd @brikis98. Can you confirm me the Docker version in CI so that I can try to reproduce it?

Also, does this error occurs in the other tests?

    TestInspect 2020-04-09T06:53:56Z command.go:158: Error: No such container: Unable to find image 'nginx:1.17-alpine' locally

The nginx image I used is the same one on the stop tests.

**Comment by GusAntoniassi** [comment]
Oh OK I think I see what is happening. I'm using the `shell.RunCommandAndGetOutput` as the container ID, and when it does not have the image ID the output is:

```
Error: No such container: Unable to find image 'nginx:1.17-alpine' locally
...
64e2a233bec6d5487b7b1ba4cb4ca3e73703eb9ef93f2935d9f22dcab0f0fa22
```
Then it tries to use that as the container ID.

@brikis98 I have managed to fix this by changing [`modules/docker/run.go`](https://github.com/gruntwork-io/terratest/blob/master/modules/docker/run.go#L74), replacing `shell.RunCommandAndGetOutputE` with `shell.RunCommandAndGetStdOutE`. 

Do you think it's safe to commit this change? Or maybe we could implement a `RunAndGetID` in `modules/docker/run.go`.

**Comment by brikis98** [comment]
Ah, good catch! I think a `RunAndGetID` would be a safer addition, at least to make the intent clearer, so issues like this are easier to debug in the future.

**Comment by GusAntoniassi** [comment]
Done @brikis98. I've also added a `docker rmi` command in the `TestInspect`, to make the local testing consistent with the CI one.

File: modules/docker/inspect_test.go [hunk 17]
```diff
@@ -0,0 +1,126 @@
...
+	// append timestamp to container name to allow running tests in parallel
+	name := "inspect-test-" + random.UniqueId()
+
+	// running the container detached to allow inspection while it is running
+	options := &RunOptions{
+		Detach: true,
+		Name:   name,
+	}
+
+	// remove the local image if it exists to force Docker to pull it
```
Comment (brikis98) [inline modules/docker/inspect_test.go:27]: Why do we want to force a `pull`?

File: modules/docker/run.go [hunk 18]
```diff
@@ -74,6 +74,28 @@ func RunE(t testing.TestingT, image string, options *RunOptions) (string, error)
 	return shell.RunCommandAndGetOutputE(t, cmd)
 }
 
+func RunAndGetID(t testing.TestingT, image string, options *RunOptions) string {
```
Comment (brikis98) [inline modules/docker/run.go:77]: Could you add comments describing what these two new methods do?

File: modules/docker/inspect_test.go (hunk 17 above)
Comment (GusAntoniassi) [reply modules/docker/inspect_test.go:27]: I didn't catch this error when developing because I already had the image downloaded and cached. 

I added this test to validate that scenario, in case something changes (either in Terratest or in Docker) so that we can have the same behavior in CI and local testing, catching it quicker. 

Do you think it would be better to document that in the comment (line 27) too?

Comment (brikis98) [reply modules/docker/inspect_test.go:27]: Well, the test _already_ failed due to this issue, so our testing was good enough to catch it. Sure, we didn't catch it until CI, but IMO, that's a better trade-off than (a) deleting a local Docker image, which would be an unexpected thing for a test to do and (b) making each test run slower.

Comment (GusAntoniassi) [reply modules/docker/inspect_test.go]: That makes sense @brikis98! I've removed the `rmi` command in [73ef44b](https://github.com/gruntwork-io/terratest/pull/492/commits/73ef44be58fe786ebfcb655297d7c06391712ad7).

**Comment by brikis98** [comment]
One test failed due to what seems to be an unrelated, transient error with APT:

```
TestPackerMultipleConcurrentAmis 2020-04-14T08:18:14Z command.go:158: 1586852294,,ui,message,    amazon-ebs: Get:20 http://archive.ubuntu.com/ubuntu xenial-backports/main Translation-en [4456 B]
TestPackerMultipleConcurrentAmis 2020-04-14T08:18:14Z command.go:158: 1586852294,,ui,message,    amazon-ebs: Get:21 http://archive.ubuntu.com/ubuntu xenial-backports/universe amd64 Packages [8064 B]
TestPackerMultipleConcurrentAmis 2020-04-14T08:18:14Z command.go:158: 1586852294,,ui,message,    amazon-ebs: Get:22 http://archive.ubuntu.com/ubuntu xenial-backports/universe Translation-en [4328 B]
TestPackerMultipleConcurrentAmis 2020-04-14T08:18:15Z command.go:158: 1586852295,,ui,message,    amazon-ebs: Reading package lists...
TestPackerMultipleConcurrentAmis 2020-04-14T08:18:15Z command.go:158: 1586852295,,ui,error,==> amazon-ebs: Splitting up /var/lib/apt/lists/archive.ubuntu.com_ubuntu_dists_xenial_InRelease into data and signature failedE: GPG error: http://archive.ubuntu.com/ubuntu xenial InRelease: Clearsigned file isn't valid%!(PACKER_COMMA) got 'NODATA' (does the network require authentication?)
TestPackerMultipleConcurrentAmis 2020-04-14T08:18:15Z command.go:158: 1586852295,,ui,say,==> amazon-ebs: Provisioning step had errors: Running the cleanup provisioner%!(PACKER_COMMA) if present...
```

I'm going to merge. Thanks for the PR!

**Comment by brikis98** [comment]
https://github.com/gruntwork-io/terratest/releases/tag/v0.26.5

**Comment by GusAntoniassi** [comment]
Thanks, @brikis98!
